### PR TITLE
217 admin role

### LINKED ISSUE #217: [217] Admin role
**Current state**
There is no role system for accounts, which means there are no admins either.

**Expected state**
There should be a way to authenticate users based on role (Admin, etc.) in the back-end.

**TO DO**
- [ ] Authentication for admin
- [ ] Authorization for admin
- [ ] Controller for admin

**Necessities**
None

**Branch**
This task will be edited in the branch [217-admin-role](/AjobK/Seaqull/tree/217-admin-role) and merged into [224-admin](/AjobK/Seaqull/tree/224-admin)

### PR DESCRIPTION
**Pull request by ryankroon00**
closes #217 

 **What is it supposed to do**
I've refactored the seeder and added seeders for the permissions and roles and created a middleware for the backend.

 **How can we test your branch**
You can test the branch by running npm run reset, it should create 4 accounts with 4 different roles, 1 default attachment, all the defined permissions and roles and a profile for each account.

 **What browser have been checked**
- [x] Chrome
- [ ] Firefox
- [ ] Opera
- [ ] Internet Explorer
- [ ] Edge
- [ ] Safari

### CODE REVIEW

File: backend/app/controllers/roleController.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Request, Response } from 'express'
+//import Role from '../entity/role'
```
Comment (MrGooss) [inline backend/app/controllers/roleController.ts]: remove comment?

File: backend/app/daos/permissionDAO.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        return permissionList
+    }
+
+    public async getPermission(id: number): Promise<Permission> {
+        const repository = await DatabaseConnector.getRepository('Permission')
+        return await repository.find({ where: { id: id } })
+    }
+}
+
+export default PermissionDAO
```
Comment (MrGooss) [inline backend/app/daos/permissionDAO.ts]: Newline end of file

File: backend/app/daos/rolePermissionDAO.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+import { RolePermission } from '../entities/rolePermission'
+
+class RolePermissionDAO {
+    public async getRolePermissionsByRole(id: number): Promise<RolePermission[]> {
+        const repository = await DatabaseConnector.getRepository('RolePermission')
+        const rolePermissions = await repository.find({ where: { role: id }, relations: ['permission'] })
+        return rolePermissions
+    }
+}
+export default RolePermissionDAO
```
Comment (MrGooss) [inline backend/app/daos/rolePermissionDAO.ts]: newline end of file

File: backend/app/entities/rolePermission.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    @ManyToOne(() => Role, role => role.id)
+    @JoinColumn({ name: 'role_id', referencedColumnName: 'id' })
+    role: Role
+
+    @ManyToOne(() => Permission, permission => permission.id)
+    @JoinColumn({ name: 'permission_id', referencedColumnName: 'id' })
+    permission: Permission
+}
+export default RolePermission
```
Comment (MrGooss) [inline backend/app/entities/rolePermission.ts]: newline end of file

File: backend/app/factories/permissionFactory.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { define } from 'typeorm-seeding'
+import Permission from '../entities/permission'
+import * as Faker from 'faker'
+
+define(Permission, (faker: typeof Faker, settings: { name: string }): Permission => {
+    const permission = new Permission()
+    permission.name = settings.name
+    return permission
+})
```
Comment (MrGooss) [inline backend/app/factories/permissionFactory.ts]: newline end of file

File: backend/app/middlewares/hasPermission.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+                hasPermission = true
+            }
+        })
+        if (!hasPermission) {
+            return res.status(401).json({ 'error': 'You don\'t have the right permisions' })
+        }
+
+        next()
+    }
+}
```
Comment (MrGooss) [inline backend/app/middlewares/hasPermission.ts]: newline end of file

File: backend/app/seeder/mainSeeder.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        const modPermisionsObjects = (await permissionSeeder(this.permissionsModerators, factory)).concat(userPermissionsObjects)
+        const adminPermsionsObjects = (await permissionSeeder(this.permissionsAdmins, factory)).concat(modPermisionsObjects)
+        const headAdminPermissionsObject = (await permissionSeeder(this.permissionsHeadAdmins, factory)).concat(adminPermsionsObjects)
+
+        rolePermissionSeeder(userPermissionsObjects, userRole, factory)
+        rolePermissionSeeder(modPermisionsObjects, modRole, factory)
+        rolePermissionSeeder(adminPermsionsObjects, adminRole, factory)
+        rolePermissionSeeder(headAdminPermissionsObject, headAdminRole, factory)
+    }
+}
```
Comment (MrGooss) [inline backend/app/seeder/mainSeeder.ts]: newline end of file

File: backend/app/seeder/seeders/accountSeeder.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import Role from '../../entities/role'
+
+module.exports = async (factory: Factory, role: Role, profilePic: Attachment) => {
+    return await factory(Account)({ role: role, profilePic: profilePic })
+        .map(async (account: Account) => {
+            await factory(Post)({ profile: account.profile }).create()
+
+            return account
+        }).createMany(1)
+}
```
Comment (MrGooss) [inline backend/app/seeder/seeders/accountSeeder.ts]: newline end of file

File: backend/app/seeder/seeders/attachmentSeeder.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+import { Factory } from 'typeorm-seeding'
+import Attachment from '../../entities/attachment'
+
+module.exports = async (factory: Factory) => {
+    return await factory(Attachment)().create()
+}
```
Comment (MrGooss) [inline backend/app/seeder/seeders/attachmentSeeder.ts]: newline end of file

File: backend/app/seeder/seeders/permissionSeeder.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+module.exports = async (permissionsArray: Permission[], factory: Factory) => {
+    const permissions = []
+
+    for (let i = 0; i < permissionsArray.length; i++) {
+        const userPermission = await factory(Permission)({ name: permissionsArray[i] }).create()
+        permissions.push(userPermission)
+    }
+    return permissions
+}
```
Comment (MrGooss) [inline backend/app/seeder/seeders/permissionSeeder.ts]: newline end of file

File: backend/app/seeder/seeders/rolePermissionSeeder.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import Permission from '../../entities/permission'
+import RolePermission from '../../entities/rolePermission'
+
+module.exports = async (permissionsArray: Permission[], role, factory) => {
+    for (let i = 0; i < permissionsArray.length; i++) {
+        await factory(RolePermission)({ role: role, permission: permissionsArray[i] }).create()
+    }
+}
```
Comment (MrGooss) [inline backend/app/seeder/seeders/rolePermissionSeeder.ts]: newline end of file

File: backend/app/seeder/seeders/roleSeeder.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+import { Factory } from 'typeorm-seeding'
+import Role from '../../entities/role'
+
+module.exports = async (factory: Factory, name: string) => {
+    return await factory(Role)({ name: name }).create()
+}
```
Comment (MrGooss) [inline backend/app/seeder/seeders/roleSeeder.ts]: newline end of file

**Review by MrGooss — CHANGES_REQUESTED** [review]

**Comment by AjobK** [comment]
image

`npm run reset` does not seem to do it for me..? No new accounts created

File: backend/app/middlewares/hasPermission.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+const rolePermissionDAO = new RolePermissionDAO
+
+module.exports = (requiredPermissionName) => {
+    return async (req, res, next) => {
+        const permissionList = await rolePermissionDAO.getRolePermissionsByRole(req.decoded.role)
+
+        let hasPermission = false
+        permissionList.forEach(permission => {
+            if(permission.permission.name == requiredPermissionName) {
```
Comment (AjobK) [inline backend/app/middlewares/hasPermission.ts]:
```suggestion
            if (permission.permission.name == requiredPermissionName) {
```

File: backend/app/middlewares/hasPermission.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+                hasPermission = true
+            }
+        })
+        if (!hasPermission) {
+            return res.status(401).json({ 'error': 'You don\'t have the right permisions' })
+        }
```
Comment (AjobK) [inline backend/app/middlewares/hasPermission.ts:19]:
```suggestion
        
        if (!hasPermission) {
            return res.status(401).json({ 'error': 'You don\'t have the right permisions' })
        }
```

File: backend/app/middlewares/isAuth.ts [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+    if (!req.cookies) return res.status(401).send('Not authorized')
+
+    const { token } = req.cookies;
+
+    if (!token) return res.status(401).send('Not authorized')
+
+    let decodedToken = ''
+
+    try{
```
Comment (AjobK) [inline backend/app/middlewares/isAuth.ts]:
```suggestion
    try {
```

File: backend/app/seeder/seeders/permissionSeeder.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
+import { Factory } from 'typeorm-seeding'
+import Permission from '../../entities/permission'
+
+module.exports = async (permissionsArray: Permission[], factory: Factory) => {
+    const permissions = []
+
+    for (let i = 0; i < permissionsArray.length; i++) {
+        const userPermission = await factory(Permission)({ name: permissionsArray[i] }).create()
+        permissions.push(userPermission)
```
Comment (AjobK) [inline backend/app/seeder/seeders/permissionSeeder.ts]:
```suggestion
        const userPermission = await factory(Permission)({ name: permissionsArray[i] }).create()
        
        permissions.push(userPermission)
```

File: backend/app/daos/roleDAO.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import DatabaseConnector from '../utils/databaseConnector'
+import { Role } from '../entities/role'
+
+class RoleDao {
+    public async getRoles(): Promise<Role[]> {
+        const repository = await DatabaseConnector.getRepository('Role');
+        const roles = await repository.find()
+        return roles
```
Comment (AjobK) [inline backend/app/daos/roleDAO.ts:9]:
```suggestion
        
        return roles
```

File: backend/app/daos/roleDAO.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public async getRoleById(id: number): Promise<Role> {
+        const repository = await DatabaseConnector.getRepository('Role');
+        const role = await repository.findOne({ id: id })
+        return role
```
Comment (AjobK) [inline backend/app/daos/roleDAO.ts:16]:
```suggestion
        
        return role
```

File: backend/app/daos/roleDAO.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public async getRoleByUser(username: string): Promise<Role> {
+        const repository = await DatabaseConnector.getRepository('Account');
+        const role = await repository.createQueryBuilder('account')
+            .leftJoinAndSelect('account.role', 'role')
+            .select(['account.role'])
+            .where('account.user_name = :user_name', { user_name: username })
+            .getRawOne()
+        return await this.getRoleById(role.role_id)
```
Comment (AjobK) [inline backend/app/daos/roleDAO.ts:27]:
```suggestion
        
        return await this.getRoleById(role.role_id)
```

File: backend/app/daos/rolePermissionDAO.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import DatabaseConnector from '../utils/databaseConnector'
+import { RolePermission } from '../entities/rolePermission'
+
+class RolePermissionDAO {
+    public async getRolePermissionsByRole(id: number): Promise<RolePermission[]> {
+        const repository = await DatabaseConnector.getRepository('RolePermission')
+        const rolePermissions = await repository.find({ where: { role: id }, relations: ['permission'] })
+        return rolePermissions
```
Comment (AjobK) [inline backend/app/daos/rolePermissionDAO.ts:9]:
```suggestion
        
        return rolePermissions
```

File: backend/app/factories/permissionFactory.ts [hunk 21]
```diff
@@ -0,0 +1,9 @@
+import { define } from 'typeorm-seeding'
+import Permission from '../entities/permission'
+import * as Faker from 'faker'
+
+define(Permission, (faker: typeof Faker, settings: { name: string }): Permission => {
+    const permission = new Permission()
+    permission.name = settings.name
+    return permission
```
Comment (AjobK) [inline backend/app/factories/permissionFactory.ts:9]:
```suggestion
    
    return permission
```

File: backend/app/factories/rolePermissionFactory.ts [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+import { Role } from '../entities/role'
+import * as Faker from 'faker'
+import RolePermission from '../entities/rolePermission'
+import Permission from '../entities/permission'
+
+define(RolePermission, (faker: typeof Faker, settings: { role: Role, permission: Permission }):RolePermission => {
+    const rolePermissionFake = new RolePermission()
+    rolePermissionFake.role = settings.role
+    rolePermissionFake.permission = settings.permission
+    return rolePermissionFake
```
Comment (AjobK) [inline backend/app/factories/rolePermissionFactory.ts:12]:
```suggestion
    
    return rolePermissionFake
```

File: backend/app/middlewares/hasPermission.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
+import RolePermissionDAO from '../daos/rolePermissionDAO'
+
+const rolePermissionDAO = new RolePermissionDAO
+
+module.exports = (requiredPermissionName) => {
+    return async (req, res, next) => {
+        const permissionList = await rolePermissionDAO.getRolePermissionsByRole(req.decoded.role)
+
+        let hasPermission = false
+        permissionList.forEach(permission => {
```
Comment (AjobK) [inline backend/app/middlewares/hasPermission.ts]:
```suggestion
        const permissionList = await rolePermissionDAO.getRolePermissionsByRole(req.decoded.role)
        let hasPermission = false
      
        permissionList.forEach(permission => {
```

File: backend/app/middlewares/hasPermission.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+            if(permission.permission.name == requiredPermissionName) {
+                hasPermission = true
+            }
+        })
+        if (!hasPermission) {
```
Comment (AjobK) [inline backend/app/middlewares/hasPermission.ts:17]:
```suggestion
       
        if (!hasPermission) {
```

File: backend/app/seeder/seeders/rolePermissionSeeder.ts [hunk 25]
```diff
@@ -0,0 +1,8 @@
+import Permission from '../../entities/permission'
+import RolePermission from '../../entities/rolePermission'
+
+module.exports = async (permissionsArray: Permission[], role, factory) => {
+    for (let i = 0; i < permissionsArray.length; i++) {
+        await factory(RolePermission)({ role: role, permission: permissionsArray[i] }).create()
```
Comment (AjobK) [inline backend/app/seeder/seeders/rolePermissionSeeder.ts]:
```suggestion
        await factory(RolePermission)({
            role: role,
            permission: permissionsArray[i]
        }).create()
```

File: backend/app/seeder/mainSeeder.ts [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+
+        const profilePic = await attachmentSeeder(factory)
+
+        await accountSeeder(factory, userRole, profilePic)
+        await accountSeeder(factory, modRole, profilePic)
+        await accountSeeder(factory, adminRole, profilePic)
+        await accountSeeder(factory, headAdminRole, profilePic)
+
+        const userPermissionsObjects = await permissionSeeder(this.permissionsUsers, factory)
+        const modPermisionsObjects = (await permissionSeeder(this.permissionsModerators, factory)).concat(userPermissionsObjects)
```
Comment (AjobK) [inline backend/app/seeder/mainSeeder.ts]:
```suggestion
        const modPermisionsObjects = (await permissionSeeder(
            this.permissionsModerators,
            factory
        )).concat(userPermissionsObjects)
```

Do this for all of them please. That way we won't have to purchase a monitor of 1080 x 8000000

File: backend/app/daos/attachmentDAO.ts [hunk 27]
```diff
@@ -1,5 +1,5 @@
-import DatabaseConnector from '../util/databaseConnector'
-import Attachment from '../entity/attachment';
+import DatabaseConnector from '../utils/databaseConnector'
+import Attachment from '../entities/attachment';
```
Comment (AjobK) [inline backend/app/daos/attachmentDAO.ts]:
```suggestion
import Attachment from '../entities/attachment'
```

Remove semicolons

File: backend/app/daos/roleDAO.ts [hunk 28]
```diff
@@ -0,0 +1,28 @@
+import DatabaseConnector from '../utils/databaseConnector'
+import { Role } from '../entities/role'
+
+class RoleDao {
+    public async getRoles(): Promise<Role[]> {
+        const repository = await DatabaseConnector.getRepository('Role');
```
Comment (AjobK) [inline backend/app/daos/roleDAO.ts]:
```suggestion
        const repository = await DatabaseConnector.getRepository('Role')
```

File: backend/app/daos/roleDAO.ts [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+        const roles = await repository.find()
+        return roles
+    }
+
+    public async getRoleById(id: number): Promise<Role> {
+        const repository = await DatabaseConnector.getRepository('Role');
```
Comment (AjobK) [inline backend/app/daos/roleDAO.ts]:
```suggestion
        const repository = await DatabaseConnector.getRepository('Role')
```

File: backend/app/daos/roleDAO.ts [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+        const role = await repository.findOne({ id: id })
+        return role
+    }
+
+    public async getRoleByUser(username: string): Promise<Role> {
+        const repository = await DatabaseConnector.getRepository('Account');
```
Comment (AjobK) [inline backend/app/daos/roleDAO.ts]:
```suggestion
        const repository = await DatabaseConnector.getRepository('Account')
```

File: backend/app/factories/accountFactory.ts [hunk 31]
```diff
@@ -1,16 +1,14 @@
 import { define, factory } from 'typeorm-seeding'
-import { Account } from '../entity/account'
-import role from '../entity/role'
+import { Account } from '../entities/account'
+import role from '../entities/role'
+import Profile from '../entities/profile';
```
Comment (AjobK) [inline backend/app/factories/accountFactory.ts]:
```suggestion
import Profile from '../entities/profile'
```

File: backend/app/middlewares/isAuth.ts [hunk 32]
```diff
@@ -0,0 +1,25 @@
+const jwt = require('jsonwebtoken')
+require('dotenv').config()
+
+const { JWT_SECRET } = process.env
+
+module.exports = (req, res, next) => {
+
+    if (!req.cookies) return res.status(401).send('Not authorized')
+
+    const { token } = req.cookies;
```
Comment (AjobK) [inline backend/app/middlewares/isAuth.ts]:
```suggestion
    const { token } = req.cookies
```

**Review by AjobK — CHANGES_REQUESTED** [review]
Please see comments

File: backend/app/daos/permissionDAO.ts [hunk 33]
```diff
@@ -0,0 +1,17 @@
...
+class PermissionDAO {
+    public async getPermissions(): Promise<Permission[]> {
+        const repository = await DatabaseConnector.getRepository('Permission')
+        const permissionList = repository.find()
+        return permissionList
+    }
+
+    public async getPermission(id: number): Promise<Permission> {
+        const repository = await DatabaseConnector.getRepository('Permission')
+        return await repository.find({ where: { id: id } })
```
Comment (AjobK) [inline backend/app/daos/permissionDAO.ts]:
```suggestion
        const repository = await DatabaseConnector.getRepository('Permission')
        
        return await repository.find({ where: { id: id } })
```

File: backend/app/daos/permissionDAO.ts [hunk 34]
```diff
@@ -0,0 +1,17 @@
+import DatabaseConnector from '../utils/databaseConnector'
+import { Permission } from '../entities/permission'
+
+class PermissionDAO {
+    public async getPermissions(): Promise<Permission[]> {
+        const repository = await DatabaseConnector.getRepository('Permission')
+        const permissionList = repository.find()
+        return permissionList
```
Comment (AjobK) [inline backend/app/daos/permissionDAO.ts]:
```suggestion
        const permissionList = repository.find()
        
        return permissionList
```

File: backend/ormconfig.example.json [hunk 35]
```diff
@@ -0,0 +1,26 @@
...
+        "app/subscriber/**/*.ts"
+    ],
+    "cli": {
+        "entitiesDir": "app/entity",
+        "migrationsDir": "app/migration",
+        "subscribersDir": "app/subscriber"
+    },
+    "seeds": ["app/seeder/mainSeeder.ts"],
+    "factories": ["app/factories/*.ts"]
+}
```
Comment (AjobK) [inline backend/ormconfig.example.json:26]: newline

**Review by AjobK — CHANGES_REQUESTED** [review]

**Pull request merged**